protocols/mp-ecdsa/src/protocols/keygen.rs

```diff
@@ -0,0 +1,202 @@
+        Ok(())
+    }
+
+    async fn process_error(
+        &self,
+        error: webb_gadget::Error,
+        _job_manager: &ProtocolWorkManager<WebbWorkManager>,
+    ) {
+        log::error!(target: "mp-ecdsa", "Error: {error:?}");
+    }
+
+    fn get_work_manager_config(&self) -> WorkManagerConfig {
+        WorkManagerConfig {
+            interval: None, // Manual polling
+            max_active_tasks: 1,
+            max_pending_tasks: 1,
+        }
+    }
+}
+
+struct MpEcdsaKeygenExtraParams {
+    i: u16,
+    t: u16,
+    n: u16,
+    job_id: JobId,
+}
+
+#[async_trait]
+impl<B, BE, KBE: KeystoreBackend, C> AsyncProtocol for MpEcdsaKeygenProtocol<B, BE, KBE, C> {
+    type AdditionalParams = MpEcdsaKeygenExtraParams;
+    async fn generate_protocol_from(
+        &self,
+        associated_block_id: <WebbWorkManager as WorkManagerInterface>::Clock,
+        associated_retry_id: <WebbWorkManager as WorkManagerInterface>::RetryID,
+        associated_session_id: <WebbWorkManager as WorkManagerInterface>::SessionID,
+        associated_task_id: <WebbWorkManager as WorkManagerInterface>::TaskID,
+        protocol_message_rx: UnboundedReceiver<GadgetProtocolMessage>,
+        additional_params: Self::AdditionalParams,
+    ) -> Result<BuiltExecutableJobWrapper, JobError> {
+        let key_store = self.client.key_store.clone();
+        let blame = self.round_blames.clone();
+
+        Ok(JobBuilder::new()
+            .protocol(async move {
+                let (i, t, n) = (
+                    additional_params.i,
+                    additional_params.t,
+                    additional_params.n,
+                );
+                let keygen = Keygen::new(i, t, n)?;
+                let (current_round_blame_tx, current_round_blame_rx) =
+                    tokio::sync::watch::channel(CurrentRoundBlame::default());
+
+                self.round_blames
+                    .write()
+                    .await
+                    .insert(associated_task_id, current_round_blame_rx);
+
+                let (tx_to_outbound, rx_async_proto) =
+                    util::create_job_manager_to_async_protocol_channel::<Keygen, _>(
+                        protocol_message_rx,
+                        associated_block_id,
+                        associated_retry_id,
+                        associated_session_id,
+                        associated_task_id,
+                        self.network.clone(),
+                    );
+
+                let state_machine_wrapper =
+                    StateMachineWrapper::new(keygen, current_round_blame_tx, self.logger.clone());
+                let local_key = round_based::AsyncProtocol::new(
+                    state_machine_wrapper,
+                    rx_async_proto,
+                    tx_to_outbound,
+                )
+                .set_watcher(StderrWatcher)
+                .run()
+                .await?;
+
+                key_store.set(additional_params.job_id, local_key).await?;
```

Comment:
This LocalKey, you should save the key by the `phase_one.job_id` this one you should use above (see my previous comment)